golem-worker-service-base/src/evaluator/mod.rs

```diff
@@ -1,24 +1,59 @@
+pub use evaluator_context::*;
+mod evaluator_context;
+mod getter;
+mod math_op_evaluator;
+mod path;
+mod pattern_match_evaluator;
+
 use golem_wasm_ast::analysis::AnalysedType;
 use golem_wasm_rpc::json::get_json_from_typed_value;
 use golem_wasm_rpc::TypeAnnotatedValue;
 
-use crate::primitive::GetPrimitive;
+use crate::expression;
+use crate::primitive::{GetPrimitive, Primitive};
 use getter::GetError;
 use getter::Getter;
 use path::Path;
 
 use crate::expression::{Expr, InnerNumber};
-use crate::merge::Merge;
 
-use crate::tokeniser::tokenizer::{MultiCharTokens, Token, Tokenizer};
+pub trait Evaluator {
+    fn evaluate(
```

Comment:
I think evaluation should take an input expression, which is nothing more than a record (`TypeAnnotatedValue`  or even more specific), together with a standard library -- i.e. functions which may be called by the expression (including, of course, worker functions, utility functions, etc.). A function can be defined by its formal definition (namespace, name, parameters and their types and names), and then an implementation, which is something the evaluator can call into, and which returns a type annotated value.

This is the way to enabling extensionality and also allowing Rib to be pulled out of this module and shipped as its own crate.

Reply:
Actually, this is too poorly typed: the input to the evaluator should be a map of identifier name to `TypedAnnotatedValue` (yes, a record is basically the same thing, but a bunch of let-bindings is clearer and more type safe).

Reply:
The input to the evaluator is actually an expression.
The type-annotated-value is global namespace details, which are simply represented using type-annotated value. 

Logically what we have is the following

```
fn evaluate(input: Expr, context: EvaluationContext) -> Result<EvaluationResult, EvaluationError>

```

Reply:
The input expression is of the form `Expr`. It is the `evaluator` that converts it into `TypeAnnotatedValue`.

`TypeAnnotatedValue` is not further evaluated, meaning the input of the current evaluator is `Expr` and output is `TypeAnnotatedValue` (currently its called EvaluationResult)

Reply:
We can pair up on this for a quick discussion once I reach home

Reply:
Current evaluationContext can be empty. As in, it can be any combination.  Users can ask the evaluator to evaluate an expression (to form a type annotated value) given optional worker info, optional request info etc. I don’t think we are ending up in representing any illegal states with this typing. But I would like to pair up with you and get more thoughts there.